### ISSUE TITLE
Feature proposal: Option to push data from clients to engeins through file system

### ISSUE THREAD

**Issue by ranr01** [ipython/ipyparallel#145]
I don't know if this would be interesting for the project but since I implemented it for my needs I figured I'll bring it up.

The problem: Sometimes it is needed to transfer large datasets generated on the client to the engines. In large clusters it becomes very inefficient to do that through the bottleneck of the controller. 

The solution (for systems with shared file system such as HPC clusters): Client saves data as dictionary to the file system and engines read data from file. 

Usage would be something like this:

```
import numpy as np
import ipyparallel

ippClient = ipyparallel.Client()
dv=ippClient.direct_view()

X=np.random.randn(100,100)
Y=np.random.randn(200,100)

fs_push(dv,dict(X=X,Y=Y))
```

Implementation can be something like this:

```
def _fs_push(filename):
    """helper method for implementing 'client.fs_push' via 'client.apply'"""
    ns = YOUR_LOAD_FUNCTION_HERE(filename)

    user_ns = globals()
    tmp = '_IP_PUSH_TMP_'
    while tmp in user_ns:
        tmp = tmp + '_'
    try:
        for name, value in ns.items():
            user_ns[tmp] = value
            exec("%s = %s" % (name, tmp), user_ns)
    finally:
        user_ns.pop(tmp, None)

def fs_push(self, ns, targets=None, track=None):
        """update remote namespace with dict `ns` through shared access file system.

        Parameters
        ----------
        ns : dict
            dict of keys with which to update engine namespace(s)
        """

        block = True
        track = track if track is not None else self.track
        targets = targets if targets is not None else self.targets

        if not isinstance(ns, dict):
            raise TypeError("Must be a dict, not %s"%type(ns))

        import os
        #make sure not overwriting any file
        tmp_fn = '_TMP_PUSH_FN_'
        while tmp_fn in os.listdir():
            tmp_fn = tmp_fn + '_'
        tmp_fn = os.getcwd() + '/' + tmp_fn

        #save ns to file
        YOUR_SAVE_FUNCTION_HERE(tmp_fn,ns)
        try:
            # engeins read file (blocking)
            ar = self._really_apply(_fs_push, args=(tmp_fn,), \
                                                block=block, track=track, targets=targets)
        finally:
            #remove file  
            os.remove(tmp_fn)
        # return result
        return ar
```

Without a shared file system one would probably need to resort to some form of p2p communications  between engines to do it efficiently. 

Hope this is useful and sorry for the trouble if it's not...

Cheers,
Ran

**Comment by minrk** [comment]
This is indeed useful! I have used a [similar approach](https://github.com/jupyter/ngcm-tutorial/blob/master/Day-3/examples/memmap.ipynb) in a tutorial, and we should move that info over to the examples in this repo.

But the gist is great: there are many ways to move data around, and different mechanism may be more appropriate, given the nature of your data and system. One nice thing about ipyparallel is that it lets you use whatever you want for out-of-band data movement. One less nice thing is that it doesn't do it for you.

**Comment by minrk** [comment]
Closed by #485
